File: backend/app/services/lobby_chat.py
"""The last few lines said in the lobby, and the number each one was given.

Lobby chat rides the lobby channel but is not a feed the way presence and the
room list are. Those two describe a *state* - who is online, which rooms are
open - that the tick rebuilds from a source of truth, diffs, and numbers with
a revision a client resyncs on. A chat line is an event: there is nothing to
rebuild it from, a one-second tick is latency a conversation feels, and a
missing line is not a fault to correct - a blocked line is deliberately never
delivered to the blocker. So a line is emitted the moment it is said, and this
module keeps only what a new arrival is shown.

`seq` numbers lines within one process so a client can put the backlog it was
handed and the lines that beat it into one order without a duplicate. It is
not a revision: a client never asks for a resync because of a gap in it, and
it starts again from nothing with the process, which is fine because every
socket starts again with the process too.

The lines themselves do not start again from nothing. Every accepted line is
also retained for thirty days (`message_retention.py`), so a restart re-seeds
the ring from the most recent retained rows before the first socket arrives:
somebody landing on the lobby after a deploy sees the same conversation they
would have seen a minute before it. That read is bounded and best-effort - a
database that does not answer leaves the ring empty and the process starting,
because chat is not the thing a deploy waits on.
"""
from __future__ import annotations

import asyncio
from collections import deque
from dataclasses import dataclass
from datetime import datetime, timezone
import logging
from typing import Iterable

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

from app.auth.bans import active_ban_filter
from app.db.models import RoomMessage, UserBan

LOBBY_CHAT_BACKLOG = 50
RESTORE_TIMEOUT_SECONDS = 10

logger = logging.getLogger(__name__)


@dataclass(frozen=True, slots=True)
class LobbyChatLine:
    seq: int
    user_id: str
    display_name: str
    name_color: str | None
    is_anonymous: bool
    text: str
    sent_at: datetime
    retained_message_id: str | None = None

    def payload(self) -> dict:
        """The wire shape, `LobbyChatMessage` in the wire protocol.

        Carries an account id for the reason `LobbyPlayer` does: there is no
        seat to resolve for somebody idling in the lobby, and a report needs a
        stable target. `retainedMessageId` is present only when retention took
        the row, exactly as room chat says it (R-MOD-08a).
        """
        payload = {
            "seq": self.seq,
            "userId": self.user_id,
            "displayName": self.display_name,
            "nameColor": self.name_color,
            "isAnonymous": self.is_anonymous,
            "text": self.text,
            "sentAt": self.sent_at.isoformat(),
        }
        if self.retained_message_id is not None:
            payload["retainedMessageId"] = self.retained_message_id
        return payload


class LobbyChatLog:
    """A bounded ring of the most recent lobby lines, oldest first."""

    def __init__(self, *, backlog: int = LOBBY_CHAT_BACKLOG) -> None:
        self._lines: deque[LobbyChatLine] = deque(maxlen=backlog)
        self._seq = 0

    @property
    def last_seq(self) -> int:
        """The number the most recent line was given; zero before any."""
        return self._seq

    def append(
        self,
        *,
        user_id: str,
        display_name: str,
        name_color: str | None,
        is_anonymous: bool,
        text: str,
        sent_at: datetime,
        retained_message_id: str | None = None,
    ) -> LobbyChatLine:
        """Number one accepted line and keep it.

        Called only once a line has passed every refusal, so a refused line
        never spends a number or shows up in anybody's backlog.
        """
        if sent_at.tzinfo is None:
            raise ValueError("sent_at must be timezone-aware")
        self._seq += 1
        line = LobbyChatLine(
            seq=self._seq,
            user_id=user_id,
            display_name=display_name,
            name_color=name_color,
            is_anonymous=is_anonymous,
            text=text,
            sent_at=sent_at,
            retained_message_id=retained_message_id,
        )
        self._lines.append(line)
        return line

    def restore(self, rows: Iterable[RoomMessage]) -> int:
        """Seed an empty ring from retained rows, oldest first.

        Only ever an empty one: a line said since the process started is
        newer than anything the database holds, and numbering the restored
        lines behind it would show them out of order. Returns how many were
        taken.
        """
        if self._lines or self._seq:
            raise RuntimeError("the lobby backlog can only be restored before use")
        count = 0
        for row in rows:
            if row.sender_user_id is None:
                continue
            self.append(
                user_id=str(row.sender_user_id),
                display_name=row.sender_display_name_snapshot,
                name_color=row.sender_name_color_snapshot,
                is_anonymous=row.sender_is_anonymous_snapshot,
                text=row.text,
                sent_at=row.created_at,
                retained_message_id=str(row.id),
            )
            count += 1
        return count

    def backlog_for(self, *, hidden_authors: Iterable[str] = ()) -> list[LobbyChatLine]:
        """What one arriving watcher is shown, minus the authors they blocked."""
        hidden = frozenset(hidden_authors)
        return [line for line in self._lines if line.user_id not in hidden]

    def authors(self) -> set[str]:
        """Who wrote what is held, for the block lookups an arrival needs."""
        return {line.user_id for line in self._lines}

    def drop_author(self, user_id: str) -> None:
        """Forget every line by one account.

        A deletion or a suspension ends the account; its name and its words
        must not keep being handed to every new arrival for the next fifty
        lines. The numbers already given stay spent.
        """
        kept = [line for line in self._lines if line.user_id != user_id]
        self._lines.clear()
        self._lines.extend(kept)


async def recent_lobby_lines(
    session: AsyncSession, *, limit: int = LOBBY_CHAT_BACKLOG, now: datetime | None = None
) -> list[RoomMessage]:
    """The most recent retained lobby lines, oldest first.

    Minus anything expired and anything said by an account that is suspended
    now - the ban sweep dropped those from the live ring, and a restart must
    not hand them back to the next fifty arrivals. A deleted account's rows
    are already gone (`auth/account_data.py`).
    """
    checked_at = now or datetime.now(timezone.utc)
    suspended = select(UserBan.user_id).where(*active_ban_filter(checked_at))
    rows = (
        await session.scalars(
            select(RoomMessage)
            .where(
                RoomMessage.audience == "lobby",
                RoomMessage.expires_at > checked_at,
                RoomMessage.sender_user_id.is_not(None),
                RoomMessage.sender_user_id.not_in(suspended),
            )
            .order_by(RoomMessage.created_at.desc(), RoomMessage.id.desc())
            .limit(limit)
        )
    ).all()
    return list(reversed(rows))


async def restore_lobby_backlog(
    log: LobbyChatLog,
    session_factory: async_sessionmaker[AsyncSession] | None,
    *,
    bound_seconds: float = RESTORE_TIMEOUT_SECONDS,
) -> int:
    """Re-seed the ring from the database at startup; never raise.

    Bounded, because this runs before the process serves anybody and a
    database that has stopped answering must not turn a deploy into an
    outage over the chat backlog. An empty ring is the same thing a first
    ever start has.
    """
    if session_factory is None:
        return 0
    try:
        async with session_factory() as session:
            rows = await asyncio.wait_for(
                recent_lobby_lines(session), timeout=bound_seconds
            )
        return log.restore(rows)
    except Exception:
        logger.warning(
            "Could not restore the lobby chat backlog; starting with none",
            exc_info=True,
        )
        return 0

File: frontend/src/lib/lobbyChat.ts
/** The lobby's chat: the lines the client holds, and the rules it applies.

Kept out of the component for the reason `lobbyPresence.ts` gives: the unit
tests run on bare `node:test`, so everything here is a pure function over
plain objects.

Chat rides the lobby channel but is not a feed of it. Presence and the room
list are *state*, rebuilt by the server on a tick and numbered with a revision
that a gap in means a resync. A chat line is an event: it goes out the moment
it is said, and a gap in its numbering is expected rather than a fault - a
line is deliberately not delivered to somebody who blocked its author. So
`seq` here is used for one thing only: putting the backlog an acknowledgement
hands over and the lines that beat it into one order without a duplicate. A
line numbered at or below what is held is one we have; nothing ever asks for a
resync because of it. */
import { formatClock, type TimeFormat } from "./clock.ts";


export interface LobbyChatLine {
  seq: number;
  userId: string;
  displayName: string;
  nameColor: string | null;
  isAnonymous: boolean;
  text: string;
  /** The server's instant, as epoch milliseconds. Rendered, never sorted by. */
  sentAt: number;
  /** Present only when retention took the row - the id a report can cite. */
  retainedMessageId?: string;
}

export interface LobbyChatState {
  /** The highest sequence number seen, held or not. Zero before any. */
  lastSeq: number;
  lines: LobbyChatLine[];
  /** Which server process numbered these lines (`chatEpoch`), so a later
  `watch_lobby` can ask only for newer ones - and only of the same process. */
  epoch: string | null;
  /** The account these lines were filtered for. Another account's blocks
  differ, so its lobby asks for everything and replaces them. */
  owner: string | null;
}

export const EMPTY_LOBBY_CHAT: LobbyChatState = {
  lastSeq: 0,
  lines: [],
  epoch: null,
  owner: null,
};

/** More than the server hands an arrival, so a long-open lobby keeps some of
what it watched go by; bounded so it never grows with the evening. */
export const MAX_HELD_LINES = 200;

/** One line, or null if the server sent something this build cannot read. */
export function parseLine(value: unknown): LobbyChatLine | null {
  if (!value || typeof value !== "object") return null;
  const row = value as Record<string, unknown>;
  if (typeof row.seq !== "number" || !Number.isInteger(row.seq) || row.seq < 1) return null;
  if (typeof row.userId !== "string" || !row.userId) return null;
  if (typeof row.displayName !== "string") return null;
  if (typeof row.text !== "string") return null;
  // Whole seconds since the epoch (#885).
  if (typeof row.sentAt !== "number" || !Number.isFinite(row.sentAt)) return null;
  const sentAt = row.sentAt * 1000;
  const line: LobbyChatLine = {
    seq: row.seq,
    userId: row.userId,
    displayName: row.displayName,
    nameColor: typeof row.nameColor === "string" ? row.nameColor : null,
    isAnonymous: row.isAnonymous === true,
    text: row.text,
    sentAt,
  };
  if (typeof row.retainedMessageId === "string" && row.retainedMessageId) {
    line.retainedMessageId = row.retainedMessageId;
  }
  return line;
}

function capped(lines: LobbyChatLine[]): LobbyChatLine[] {
  return lines.length > MAX_HELD_LINES ? lines.slice(lines.length - MAX_HELD_LINES) : lines;
}

/** Append one line, unless it is one we hold or cannot read.

Returns the state it was given when there is nothing to do, so a store can
tell a no-op from a change by reference. */
export function applyChatLine(state: LobbyChatState, payload: unknown): LobbyChatState {
  return append(state, parseLine(payload));
}

function append(state: LobbyChatState, line: LobbyChatLine | null): LobbyChatState {
  if (!line || line.seq <= state.lastSeq) return state;
  return { ...state, lastSeq: line.seq, lines: capped([...state.lines, line]) };
}

function parseBacklog(payload: unknown): {
  lines: LobbyChatLine[];
  chatSeq: number;
  epoch: string | null;
} {
  if (!payload || typeof payload !== "object") return { lines: [], chatSeq: 0, epoch: null };
  const answer = payload as Record<string, unknown>;
  const lines = Array.isArray(answer.chat)
    ? answer.chat.map(parseLine).filter((line): line is LobbyChatLine => line !== null)
    : [];
  lines.sort((a, b) => a.seq - b.seq);
  const chatSeq =
    typeof answer.chatSeq === "number" && Number.isInteger(answer.chatSeq) && answer.chatSeq >= 0
      ? answer.chatSeq
      : 0;
  const epoch = typeof answer.chatEpoch === "string" && answer.chatEpoch ? answer.chatEpoch : null;
  return { lines, chatSeq, epoch };
}

/** What a `watch_lobby` sends about the chat it already holds (#885).

The last number seen and the numbering it belongs to, so the server sends only
newer lines - on a resync, on a reconnect to the same process, and when a
visitor naming themselves re-handshakes the socket. Nothing when there is
nothing to resume, or when the lines were filtered for another account. */
export function chatResumeRequest(
  state: LobbyChatState,
  owner: string | null,
): { chatSince: number; chatEpoch: string } | Record<string, never> {
  if (!state.epoch || state.lastSeq === 0 || state.owner !== owner) return {};
  return { chatSince: state.lastSeq, chatEpoch: state.epoch };
}

/** Take the backlog a `watch_lobby` acknowledgement carries, for *owner*.

Merged when it continues what is held - the same process, and lines filtered
for the same account - so a lobby left open all evening keeps what it watched
go by rather than being cut back to the fifty the server keeps, and a resumed
answer, which holds only the newer lines, adds to them. Replaced otherwise: the
numbers held belong to a sequence that no longer exists, or to somebody else's
blocks, and what the server hands over is all there is. */
export function applyChatBacklog(
  state: LobbyChatState,
  payload: unknown,
  owner: string | null,
): LobbyChatState {
  const { lines, chatSeq, epoch } = parseBacklog(payload);
  if (!epoch || epoch !== state.epoch || owner !== state.owner) {
    const last = lines.length ? lines[lines.length - 1].seq : 0;
    return { lastSeq: Math.max(chatSeq, last), lines: capped(lines), epoch, owner };
  }
  let next = state;
  for (const line of lines) next = append(next, line);
  if (chatSeq > next.lastSeq) next = { ...next, lastSeq: chatSeq };
  return next;
}

/** Whether this viewer may report this line from the lobby.

A line is reported over REST, citing its retained row, because the lobby has
no seat for the socket route to resolve. So there has to be a row to cite: a
line retention withheld is shown like any other and simply offers nothing,
rather than a dialog that would be refused on sending. The other two rules
are the room's: nobody reports themselves, and a guest is offered no control
because a report a moderator cannot follow up on helps nobody (R-MOD-06). */
export function reportableLine(
  line: LobbyChatLine,
  viewer: { id: string; isAnonymous: boolean } | null | undefined,
): boolean {
  if (!viewer || viewer.isAnonymous) return false;
  if (line.userId === viewer.id) return false;
  return Boolean(line.retainedMessageId);
}

const MINUTE_MS = 60_000;
const HOUR_MS = 60 * MINUTE_MS;
const DAY_MS = 24 * HOUR_MS;

function localMidnight(at: number): number {
  const date = new Date(at);
  return new Date(date.getFullYear(), date.getMonth(), date.getDate()).getTime();
}

/** The small label beside a line: how fresh it is, at a glance.

Fresh lines say how long ago, because that is the question - is anybody
still here? A line from earlier today says when, because "six hours ago"
is arithmetic the reader would do anyway. Older than that, the day is all
that matters. A clock behind the server's reads as "now" rather than as a
line from the future. */
export function chatTimeLabel(
  sentAt: number,
  now: number,
  timeFormat: TimeFormat = "system",
): string {
  const age = now - sentAt;
  if (age < MINUTE_MS) return "now";
  if (age < HOUR_MS) return `${Math.floor(age / MINUTE_MS)}m`;
  const today = localMidnight(now);
  const thatDay = localMidnight(sentAt);
  if (thatDay === today) {
    return formatClock(new Date(sentAt), timeFormat);
  }
  const days = Math.round((today - thatDay) / DAY_MS);
  return days <= 1 ? "yesterday" : `${days}d`;
}
